topostats/tracing/ordered_tracing.py

```diff
@@ -0,0 +1,729 @@
+
+        Returns
+        -------
+        npt.NDArray
+            2xN coordinate array without the previous segment coordinates.
+        """
+        # Convert arrays to tuples
+        curr_segment_tuples = [tuple(row) for row in current_segment]
+        prev_segment_tuples = [tuple(row) for row in prev_segment]
+        # Find unique rows
+        unique_rows = list(set(curr_segment_tuples) - set(prev_segment_tuples))
+        # Remove duplicate rows from array1
+        return np.array([row for row in curr_segment_tuples if tuple(row) in unique_rows])
+
+    @staticmethod
+    def order_from_end(last_segment_coord: npt.NDArray, current_segment: npt.NDArray) -> npt.NDArray:
+        """
+        Order the current segment to follow from the end of the previous one.
+
+        Parameters
+        ----------
+        last_segment_coord : npt.NDArray
+            X and Y coordinates of the end of the last segment.
+        current_segment : npt.NDArray
+            A 2xN array of coordinates of the current segment to order.
+
+        Returns
+        -------
+        npt.NDArray
+            The current segment orientated to follow on from the last.
+        """
+        start_xy = current_segment[0]
+        dist = np.sum((start_xy - last_segment_coord) ** 2) ** 0.5
+        if dist <= np.sqrt(2):
+            return current_segment
+        return current_segment[::-1]
+
+    def get_over_under_img(self, coord_trace: list, fwhms: list, crossing_coords: list) -> npt.NDArray:
+        """
+        Obtain a labeled image according to the main trace (=1), under (=2), over (=3).
+
+        Parameters
+        ----------
+        coord_trace : list
+            Ordered coordinate trace of each molecule.
+        fwhms : list
+            List of FWHMs for each crossing in the trace.
+        crossing_coords : list
+            The crossing coordinates of each branch crossing.
+
+        Returns
+        -------
+        npt.NDArray
+            2D crossing order labeled image.
+        """
+        # put down traces
+        img = np.zeros_like(self.skeleton)
+        for _, coords in enumerate(coord_trace):
+            temp_img = np.zeros_like(img)
+            temp_img[coords[:, 0], coords[:, 1]] = 1
+            # temp_img = binary_dilation(temp_img)
+            img[temp_img != 0] = 1  # mol_no + 1
+        lower_idxs, upper_idxs = self.get_trace_idxs(fwhms)
+
+        # plots over/unders
+        for i, type_idxs in enumerate([lower_idxs, upper_idxs]):
+            for crossing, type_idx in zip(crossing_coords, type_idxs):
+                temp_img = np.zeros_like(img)
+                cross_coords = crossing[type_idx]
+                # decide which val
+                matching_coords = np.array([])
+                c = 0
+                # get overlaps between segment coords and crossing under coords
+                for cross_coord in cross_coords:
+                    c += ((coord_trace[0] == cross_coord).sum(axis=1) == 2).sum()
```

Comment:
Another good spot, so this was code which was split out from a function which did both get_over_under_img and get_mols_img, thus the for loop and associated vars are leftover and only used in the other func.

But in the other func, this is used to see which segment is over / under the other and plot them on the image according to their stacking order. For this, only the first coordinate is needed as this will be away from the crossing point and thus not present in both. All the coords could be used but becomes and issue when the branches are short and this difference not present. i.e. green vs pink below but only looking at the start coord.
<img width="531" alt="Screenshot 2024-08-14 at 22 50 48" src="https://github.com/user-attachments/assets/256e0ce4-aa3a-4ab0-aaa2-0a1dc3b66185">

Then this is used to plot whichever molecule segment is on top. i.e. green over blue on the right, and blue over green on the left.
<img width="583" alt="Screenshot 2024-08-14 at 22 51 18" src="https://github.com/user-attachments/assets/5546bcff-eb5a-4a13-849e-3e285d000abc">